src/basic/conf-files.c

```diff
@@ -104,27 +102,23 @@ static int files_add(
                                 continue;
                         }
 
-                if (flags & CONF_FILES_BASENAME) {
-                        p = strdup(de->d_name);
-                        if (!p)
-                                return -ENOMEM;
-
-                        key = p;
-                } else {
+                if (!(flags & CONF_FILES_BASENAME)) {
                         p = path_join(dirpath, de->d_name);
                         if (!p)
                                 return -ENOMEM;
-
-                        key = basename(p);
                 }
 
-                r = hashmap_put(h, key, p);
-                if (r < 0) {
-                        free(p);
-                        return log_debug_errno(r, "Failed to add item to hashmap: %m");
-                }
+                n = strdup(de->d_name);
+                if (!n)
+                        return -ENOMEM;
 
+                r = hashmap_ensure_put(h, p ? &string_hash_ops_free_free : &string_hash_ops_free, n, p ?: n);
+                if (r < 0)
+                        return r;
```

Comment:
it hink it would be more readable to do it like this:

```c
n = strdup(de->d_name);
if (!n)
        return -ENOMEM;

if (flags & CONF_FILES_BASENAME) 
        r = hashmap_ensure_put(h, &string_hash_ops_free, n, n);
else {
        p = path_join(dirpath, de->d_name);
        if (!p)
                return -ENOMEM;

        r = hashmap_ensure_put(h, &string_hash_ops_free_free, n, p);
}
if (r < 0)
        return r;
```